rust/automerge-wasm/src/export_cache.rs

```diff
@@ -339,77 +418,111 @@ impl<'a> ExportCache<'a> {
+        datatype: Datatype,
+        user_data: &JsValue,
+        raw: ObjFormat,
+    ) -> JsValue {
+        let mut bytes = obj.map(|o| o.to_bytes()).unwrap_or_default();
+        bytes.push(datatype as u8);
+        let state = Array::new();
+        state.push(&Uint8Array::from(bytes.as_slice()));
+        state.push(user_data);
+        match &raw {
+            ObjFormat::CustomObject(o) => {
+                state.push(o);
+            }
+            ObjFormat::CustomScalar(s) => {
+                state.push(s);
+            }
+            _ => {}
+        }
+        JsValue::from(state)
     }
 
-    pub(crate) fn set_raw_data(&self, obj: &Object, value: &JsValue) -> Result<(), error::Export> {
-        self.set_value(obj, &self.raw_data_sym, value)
+    pub(crate) fn get_obj_metadata(&self, obj: &Object) -> Option<ObjMetadata> {
+        let meta = self.metadata.get(obj);
+        Self::decode_obj_metadata(&meta)
     }
 
-    pub(crate) fn get_raw_object(&self, obj: &JsValue) -> Result<JsValue, error::GetProp> {
-        self.get_value(obj, &self.raw_obj_sym)
+    pub(crate) fn decode_obj_metadata(state: &JsValue) -> Option<ObjMetadata> {
+        let array = state.clone().dyn_into::<Array>().ok()?;
+        let mut raw = ObjFormat::Normal;
+        if array.length() == 3 {
+            let inner = array.get(2);
+            if inner.is_object() {
+                raw = ObjFormat::CustomObject(inner.dyn_into::<Object>().ok()?);
+            } else {
+                raw = ObjFormat::CustomScalar(inner);
+            }
+        }
+        let user_data = array.get(1);
+        //let mut bytes = array.pop().dyn_into::<Uint8Array>().ok()?.to_vec();
+        let bytes = array.get(0).dyn_into::<Uint8Array>();
+        let mut bytes = bytes.ok()?.to_vec();
+        let datatype = bytes.pop()?;
+        let datatype = Datatype::try_from(datatype).ok()?;
+        let obj = ObjId::try_from(bytes.as_slice()).ok();
+        Some(ObjMetadata {
+            obj,
+            datatype,
+            user_data,
+            raw,
+        })
     }
 
-    pub(crate) fn set_raw_object(
+    pub(crate) fn set_obj_metadata(
         &self,
-        obj: &Object,
-        value: &JsValue,
-    ) -> Result<(), error::Export> {
-        self.set_value(obj, &self.raw_obj_sym, value)
+        js_obj: &Object,
+        id: Option<&ObjId>,
+        datatype: Datatype,
+        raw: ObjFormat,
+    ) {
+        let encoded = Self::encode_obj_metadata(id, datatype, &self.user_data, raw);
+        self.metadata.set(js_obj, &encoded);
     }
 
-    pub(crate) fn get_datatype(&self, obj: &JsValue) -> Result<JsValue, error::GetProp> {
-        self.get_value(obj, &self.datatype_sym)
+    pub(crate) fn freeze_objects(&self) {
+        for cow in self.objs.values() {
+            Object::freeze(&cow.outer);
+        }
     }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct COWJsObj {
+    pub(crate) metadata: ObjMetadata,
+    pub(crate) outer: Object,
```

Comment:
I think the "inner" and "outer" terminology here is a bit confusing now that this is called COWJsObj. Do I understand it correctly that "outer" is the value we will be passing back to the user whilest "inner" is the unwrapped value (i.e. we need to combine it with the metadata.datatype to understand it)?